src/runtime/strategies/map-slots.ts

```diff
@@ -95,50 +120,58 @@ export class MapSlots extends Strategy {
-    if (!MapSlots.specMatch(slotConnection, slot)) {
+  static slotMatches(particle: Particle, slotSpec: SlotSpec, slot) {
+    if (!MapSlots.specMatch(slotSpec, slot.spec)) {
       return false;
     }
 
-    if (!MapSlots.tagsOrNameMatch(slotConnection, slot)) {
+    const potentialSlotConn = particle.getSlotConnectionBySpec(slotSpec);
+    if (!MapSlots.tagsOrNameMatch(slotSpec, slot.spec, potentialSlotConn, slot)) {
       return false;
     }
 
     // Match handles of the provided slot with the slot-connection particle's handles.
-    if (!MapSlots.handlesMatch(slotConnection, slot)) {
+    if (!MapSlots.handlesMatch(particle, slot)) {
       return false;
     }
     return true;
   }
 
-  static specMatch(slotConnection, slot) {
-    return slotConnection.slotSpec && // if there's no slotSpec, this is just a slot constraint on a verb
-          slot.spec && // if there is no spec on the slot, it is a hosted slot in the inner arc
-          slotConnection.slotSpec.isSet === slot.spec.isSet;
-  }
-
-  // Returns true, if the slot connection's tags intersection with slot's tags is nonempty.
-  // TODO: replace with generic tag matcher
-  static tagsOrNameMatch(slotConnection, slot) {
-    const consumeConnTags = [].concat(slotConnection.slotSpec.tags || [], slotConnection.tags, slotConnection.targetSlot ? slotConnection.targetSlot.tags : []);
-    const slotTags = new Set([].concat(slot.tags, slot.spec.tags || [], [slot.name]));
-    // Consume connection tags aren't empty and intersection with the slot isn't empty.
-    if (consumeConnTags.length > 0 && consumeConnTags.some(t => slotTags.has(t))) {
-      return true;
-    }
-    // For backward compatibility support explicit slot names matching.
-    return (slotConnection.name === slot.name);
+  static specMatch(slotSpec, providedSlotSpec) {
+    return slotSpec && // if there's no slotSpec, this is just a slot constraint on a verb
+            providedSlotSpec &&      
+            slotSpec.isSet === providedSlotSpec.isSet;
   }
 
   // Returns true, if the providing slot handle restrictions are satisfied by the consuming slot connection.
   // TODO: should we move some of this logic to the recipe? Or type matching?
-  static handlesMatch(slotConnection: SlotConnection, slot) {
+  static handlesMatch(particle: Particle, slot) {
     if (slot.handles.length === 0) {
       return true; // slot is not limited to specific handles
     }
-    return Object.values(slotConnection.particle.connections).find(handleConn => {
+    return Object.values(particle.connections).find(handleConn => {
       return slot.handles.includes(handleConn.handle) ||
               (handleConn.handle && handleConn.handle.id && slot.handles.map(sh => sh.id).includes(handleConn.handle.id));
     });
   }
+
+  static tagsOrNameMatch(consumeSlotSpec: SlotSpec, provideSlotSpec: ProvidedSlotSpec,  consumeSlotConn: SlotConnection = undefined,  provideSlot: Slot = undefined) {
```

Comment:
is lint happy with double space after `,`?

Reply:
I didn't catch it because lint didn't complain. All fixed now